controllers/product.go

```diff
@@ -0,0 +1,178 @@
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"err": "there is no user with id " + (string)(input.UserID)})
+		return
+	}
+
+	email := c.Request.Header.Get("email")
+	if email != user.Email {
+		c.JSON(http.StatusUnauthorized, gin.H{"error": "You are not authorized to be here"})
+		c.Abort()
+		return
+	}
+
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	// Extract image from the form
+	r := c.Request
+	r.ParseMultipartForm(10 << 20)
+	file, _, err := r.FormFile("image")
+	if err != nil {
+		fmt.Println("Error Retrieving the File")
+		fmt.Println(err)
+		return
+	}
+	defer file.Close()
+
+	// Store image in the server side
+	tempFile, err := ioutil.TempFile("res", "upload-*.png")
+	if err != nil {
+		fmt.Println(err)
+	}
+	defer tempFile.Close()
+
+	fileBytes, err := ioutil.ReadAll(file)
+	if err != nil {
+		fmt.Println(err)
+	}
+	// write this byte array to our temporary file
+	tempFile.Write(fileBytes)
+
+	// upload the image to cloudinary cloud
+	img_url, err := UploadProductImage(tempFile.Name())
+
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	// remove the image from the server side after uploading it
+	e := os.Remove(tempFile.Name())
+	if e != nil {
+		log.Fatal(e)
+	}
+
+	// Store the new product in the database
+	product := models.Product{UserID: input.UserID, Title: input.Title, Content: input.Content, Price: input.Price, ImageURL: img_url}
+	db.DB.Create(&product)
+
+	c.JSON(http.StatusOK, gin.H{"data": product})
+}
+
+func (ctrl ProductController) GetAll(c *gin.Context) {
+	var products []models.Product
+	db.DB.Find(&products)
+
+	c.IndentedJSON(http.StatusOK, gin.H{"data": products})
+}
+
+func (ctrl ProductController) GetOne(c *gin.Context) {
+
+	id := c.Param("id")
+
+	getID, err := strconv.ParseInt(id, 10, 64)
+	if getID == 0 || err != nil {
+		c.AbortWithStatusJSON(http.StatusNotFound, gin.H{"Message": "Invalid parameter"})
+		return
+	}
+
+	var product models.Product
+
+	if db.DB.Find(&product, "id=?", getID).RecordNotFound() {
+		c.IndentedJSON(http.StatusOK, gin.H{"msg": "there is no product with id " + id})
+		return
+	}
+
+	c.IndentedJSON(http.StatusOK, gin.H{"data": product})
+}
+
+func (ctrl ProductController) DeleteOne(c *gin.Context) {
+
+	id := c.Param("id")
+
+	getID, err := strconv.ParseInt(id, 10, 64)
+	if getID == 0 || err != nil {
+		c.AbortWithStatusJSON(http.StatusNotFound, gin.H{"Message": "Invalid parameter"})
+		return
+	}
+
+	var user models.User
+	var product models.Product
+	if db.DB.Find(&product, "id=?", getID).RecordNotFound() {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"err": "there is no product with id = " + strconv.Itoa((int)(getID))})
+		return
+	}
+
+	if db.DB.Find(&user, "id=?", product.UserID).RecordNotFound() {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"err": "there is no user with id " + strconv.Itoa((int)(product.UserID))})
```

Comment:
The user knows the id, we don't have to return it again. let's stick to the standard response
```
{
  success: false, 
  error: '404 Not Found'
}
```